rlwe/encryptor.go

```diff
@@ -0,0 +1,509 @@
+	encryptor.encrypt(plaintext, ciphertext, false)
+
+	return ciphertext
+}
+
+// EncryptNTTNew encrypts the input Plaintext using the stored key and returns
+// the result on a newly created Ciphertext.
+//
+// encrypt with pk: ciphertext = [pk[0]*u + m + e_0, pk[1]*u + e_1]
+// encrypt with sk: ciphertext = [-a*sk + m + e, a]
+func (encryptor *pkEncryptor) EncryptNTTNew(plaintext *Plaintext) *Element {
+
+	if encryptor.baseconverter == nil {
+		panic("Cannot EncryptNew : modulus P is empty -> use instead EncryptFastNew")
+	}
+
+	ciphertext := NewElement(encryptor.params, 1, plaintext.Level())
+	ciphertext.Value[0].IsNTT = true
+	ciphertext.Value[1].IsNTT = true
+	encryptor.encrypt(plaintext, ciphertext, false)
+
+	return ciphertext
+}
+
+func (encryptor *pkEncryptor) Encrypt(plaintext *Plaintext, ciphertext *Element) {
+
+	if encryptor.baseconverter == nil {
+		panic("Cannot Encrypt : modulus P is empty -> use instead EncryptFast")
+	}
+
+	encryptor.encrypt(plaintext, ciphertext, false)
+}
+
+func (encryptor *pkEncryptor) EncryptFastNew(plaintext *Plaintext) *Element {
+	ciphertext := NewElement(encryptor.params, 1, plaintext.Level())
+	encryptor.encrypt(plaintext, ciphertext, true)
+
+	return ciphertext
+}
+
+func (encryptor *pkEncryptor) EncryptFastNTTNew(plaintext *Plaintext) *Element {
+	ciphertext := NewElement(encryptor.params, 1, plaintext.Level())
+	ciphertext.Value[0].IsNTT = true
+	ciphertext.Value[1].IsNTT = true
+	encryptor.encrypt(plaintext, ciphertext, true)
+
+	return ciphertext
+}
+
+func (encryptor *pkEncryptor) EncryptFast(plaintext *Plaintext, ciphertext *Element) {
+	encryptor.encrypt(plaintext, ciphertext, true)
+}
+
+func (encryptor *pkEncryptor) EncryptFromCRP(plaintext *Plaintext, ciphertext *Element, crp *ring.Poly) {
+	panic("Cannot encrypt with CRP using an encryptor created with the public-key")
+}
+
+func (encryptor *pkEncryptor) EncryptFromCRPNew(plaintext *Plaintext, crp *ring.Poly) *Element {
+	panic("Cannot encrypt with CRP using an encryptor created with the public-key")
+}
+
+func (encryptor *pkEncryptor) EncryptFromCRPNTTNew(plaintext *Plaintext, crp *ring.Poly) *Element {
+	panic("Cannot encrypt with CRP using an encryptor created with the public-key")
+}
+
+// Encrypt encrypts the input Plaintext using the stored key, and returns the result
+// on the receiver Element.
+//
+// encrypt with pk: ciphertext = [pk[0]*u + m + e_0, pk[1]*u + e_1]
+// encrypt with sk: ciphertext = [-a*sk + m + e, a]
+func (encryptor *pkEncryptor) encrypt(plaintext *Plaintext, ciphertext *Element, fast bool) {
+
+	lvl := utils.MinInt(plaintext.Level(), ciphertext.Level())
+
+	poolQ0 := encryptor.poolQ[0]
+	poolP0 := encryptor.poolP[0]
+	poolP1 := encryptor.poolP[1]
+	poolP2 := encryptor.poolP[2]
+
+	// We sample a R-WLE instance (encryption of zero) over the extended ring (ciphertext ring + special prime)
+
+	ringQ := encryptor.ringQ
+
+	ciphertextNTT := ciphertext.Value[0].IsNTT
+
+	if fast {
```

Comment:
I don't know why we changed from having two different functions. It would look much better to have both a `encrypt` and `encryptFast` methods.

Reply:
fixed